tests/storage/migration.go

```diff
@@ -0,0 +1,320 @@
+			Expect(vm.Spec.Template.Spec.Volumes[0].VolumeSource.DataVolume.Name).To(Equal(name))
+
+		}
+
+		BeforeEach(func() {
+			ns = testsuite.GetTestNamespace(nil)
+			destPVC = "dest-" + rand.String(5)
+
+		})
+
+		DescribeTable("should migrate the source volume from a source DV to a destination PVC", func(mode string) {
+			vm := createVM(createDV())
+			// Create dest PVC
+			switch mode {
+			case fsPVC:
+				libstorage.CreateFSPVC(destPVC, ns, size, nil)
+			case blockPVC:
+				libstorage.CreateBlockPVC(destPVC, ns, size)
+			default:
+				Fail("Unrecognized mode")
+			}
+			By("Update volumes")
+			updateVMWithPVC(vm.Name, destPVC)
+			Eventually(func() bool {
+				vmi, err := virtClient.VirtualMachineInstance(ns).Get(context.Background(), vm.Name,
+					metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				claim := storagetypes.PVCNameFromVirtVolume(&vmi.Spec.Volumes[0])
+				return claim == destPVC
+			}, 120*time.Second, time.Second).Should(BeTrue())
+			waitForMigrationToSucceed(vm.Name, ns)
+		},
+			Entry("to a filesystem volume", fsPVC),
+			Entry("to a block volume", blockPVC),
+		)
+
+		It("should migrate the source volume from a source DV to a destination DV", func() {
+			vm := createVM(createDV())
+			destDV := createBlankDV()
+			By("Update volumes")
+			updateVMWithDV(vm.Name, destDV.Name)
+			Eventually(func() bool {
+				vmi, err := virtClient.VirtualMachineInstance(ns).Get(context.Background(), vm.Name,
+					metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				claim := storagetypes.PVCNameFromVirtVolume(&vmi.Spec.Volumes[0])
+				return claim == destDV.Name
+			}, 120*time.Second, time.Second).Should(BeTrue())
+			waitForMigrationToSucceed(vm.Name, ns)
+		})
+
+		It("should cancel the migration by the reverting to the source volume", func() {
+			dv := createDV()
+			vm := createVM(dv)
+			// Create dest PVC
+			createUnschedulablePVC(destPVC, ns, size)
+			By("Update volumes")
+			updateVMWithPVC(vm.Name, destPVC)
+			waitMigrationToExist(vm.Name, ns)
+			waitVMIToHaveVolumeChangeCond(vm.Name, ns)
+			By("Cancel the volume migration")
+			updateVMWithPVC(vm.Name, dv.Name)
+			// After the volume migration abortion the VMI should have:
+			// 1. the source volume restored
+			// 2. condition VolumesChange set to false
+			Eventually(func() bool {
+				vmi, err := virtClient.VirtualMachineInstance(ns).Get(context.Background(), vm.Name,
+					metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				claim := storagetypes.PVCNameFromVirtVolume(&vmi.Spec.Volumes[0])
+				if claim != dv.Name {
+					return false
+				}
+				conditionManager := controller.NewVirtualMachineInstanceConditionManager()
+				c := conditionManager.GetCondition(vmi, virtv1.VirtualMachineInstanceVolumesChange)
+				if c == nil {
+					return false
+				}
+				return c.Status == k8sv1.ConditionFalse
+			}, 120*time.Second, time.Second).Should(BeTrue())
+			waitMigrationToNotExist(vm.Name, ns)
+		})
+	})
```

Comment:
Test(s) with VMs that have multiple volumes, migrating a subset may be helpful

Reply:
Done now I have included a test with a VM with a containerdisk and a PVC